packages/runtime/test/integration/global-opcodebudget.ts

```
@@ -0,0 +1,141 @@
+		lsig = runtime.createLsigAccount(getProgram("lsig.teal"), []);
+
+		runtime.fundLsig(john.account, lsig.address(), 1e6);
+
+		txnParam = {
+			type: types.TransactionType.TransferAlgo,
+			sign: types.SignType.SecretKey,
+			fromAccount: john.account,
+			toAccountAddr: getApplicationAddress(appID),
+			amountMicroAlgos: 1e6,
+			payFlags: {},
+		};
+		runtime.executeTx([txnParam]);
+	});
+
+	it("Should pass signature logic", function () {
+		txnParam = {
+			type: types.TransactionType.TransferAlgo,
+			sign: types.SignType.LogicSignature,
+			lsig: lsig,
+			fromAccountAddr: lsig.address(),
+			toAccountAddr: john.address,
+			amountMicroAlgos: 0n,
+			payFlags: {
+				totalFee: 1000,
+			},
+		};
+
+		runtime.executeTx([txnParam]);
+		assert.doesNotThrow(() => runtime.executeTx([txnParam]));
+	});
+
+	describe("Test on application", () => {
+		it("call application with single tx", () => {
+			txnParam = {
+				type: types.TransactionType.CallApp,
+				sign: types.SignType.SecretKey,
+				fromAccount: john.account,
+				appID: appID,
+				appArgs: ["str:normal_tx"],
+				payFlags: {
+					totalFee: 1000,
+				},
+			};
+
+			const receipts = runtime.executeTx([txnParam]);
+			if (receipts[0].logs) {
+				assert.deepEqual(
+					receipts[0].logs[0],
+					new TextDecoder().decode(parsing.uint64ToBigEndian(693n))
+				);
+			} else {
+				assert.isTrue("false");
+			}
+		});
+
+		it("call application with group tx", () => {
+			txnParam = {
+				type: types.TransactionType.CallApp,
+				sign: types.SignType.SecretKey,
+				fromAccount: john.account,
+				appID: appID,
+				appArgs: ["str:normal_tx"],
+				payFlags: {
+					totalFee: 1000,
+				},
+			};
+
+			const buffTx = {
+				...txnParam,
+			};
+			const receipts = runtime.executeTx([txnParam, buffTx]);
+			if (receipts[0].logs) {
+				assert.deepEqual(
+					receipts[0].logs[0],
+					new TextDecoder().decode(parsing.uint64ToBigEndian(1393n))
+				);
+			} else {
+				assert.fail("log is undefined");
+			}
```

I don't think this design is correct. No need to put in `if {} else { .. }`. Assert should fail if logs is undefined.

---

I think so 👍🏻